core/src/test/java/org/apache/iceberg/TestRowLineageMetadata.java

```diff
@@ -0,0 +1,339 @@
+  public void cleanup() {
+    TestTables.clearTables();
+  }
+
+  @ParameterizedTest
+  @FieldSource("org.apache.iceberg.TestHelpers#ALL_VERSIONS")
+  public void testRowLineageSupported(int formatVersion) {
+    if (formatVersion == TableMetadata.MIN_FORMAT_VERSION_ROW_LINEAGE) {
+      assertThat(builderFor(formatVersion)).isNotNull();
+    } else {
+      IllegalArgumentException notSupported =
+          assertThrows(
+              IllegalArgumentException.class,
+              () -> TableMetadata.buildFromEmpty(formatVersion).enableRowLineage());
+      assertThat(notSupported.getMessage()).contains("Cannot use row lineage");
+    }
+  }
+
+  @ParameterizedTest
+  @FieldSource("org.apache.iceberg.TestHelpers#ALL_VERSIONS")
+  public void testSnapshotAddition(int formatVersion) {
+    assumeTrue(formatVersion >= TableMetadata.MIN_FORMAT_VERSION_ROW_LINEAGE);
+
+    Long newRows = 30L;
+
+    TableMetadata base = baseMetadata(formatVersion);
+
+    Snapshot addRows =
+        new BaseSnapshot(
+            0, 1, null, 0, DataOperations.APPEND, null, 1, "foo", base.lastRowId(), newRows);
+
+    TableMetadata firstAddition = TableMetadata.buildFrom(base).addSnapshot(addRows).build();
+
+    assertThat(firstAddition.lastRowId()).isEqualTo(newRows);
+
+    Snapshot addMoreRows =
+        new BaseSnapshot(
+            1, 2, 1L, 0, DataOperations.APPEND, null, 1, "foo", firstAddition.lastRowId(), newRows);
+
+    TableMetadata secondAddition =
+        TableMetadata.buildFrom(firstAddition).addSnapshot(addMoreRows).build();
+
+    assertThat(secondAddition.lastRowId()).isEqualTo(newRows * 2);
+  }
+
+  @ParameterizedTest
+  @FieldSource("org.apache.iceberg.TestHelpers#ALL_VERSIONS")
+  public void testInvalidSnapshotAddition(int formatVersion) {
+    assumeTrue(formatVersion >= TableMetadata.MIN_FORMAT_VERSION_ROW_LINEAGE);
+
+    Long newRows = 30L;
+
+    TableMetadata base = baseMetadata(formatVersion);
+
+    Snapshot invalidLastRow =
+        new BaseSnapshot(
+            0, 1, null, 0, DataOperations.APPEND, null, 1, "foo", base.lastRowId() - 3, newRows);
+
+    ValidationException invalidLastRowId =
+        assertThrows(
+            ValidationException.class,
+            () -> TableMetadata.buildFrom(base).addSnapshot(invalidLastRow));
+    assertThat(invalidLastRowId.getMessage()).contains("Cannot add a snapshot whose first-row-id");
+
+    Snapshot invalidNewRows =
+        new BaseSnapshot(
+            0, 1, null, 0, DataOperations.APPEND, null, 1, "foo", base.lastRowId(), null);
+
+    ValidationException nullNewRows =
+        assertThrows(
+            ValidationException.class,
+            () -> TableMetadata.buildFrom(base).addSnapshot(invalidNewRows));
+    assertThat(nullNewRows.getMessage())
+        .contains(
+            "Cannot add a snapshot with a null `addedRows` field when row lineage is enabled");
+  }
+
+  private AtomicInteger fileNum = new AtomicInteger(0);
+
+  private DataFile fileWithRows(int numRows) {
```

Comment:
```suggestion
  private DataFile fileWithRows(long numRows) {
```

nit: you might want to move this method to the top/bottom with the `fileNum` field, as otherwise it seems a bit lost between test methods